### ISSUE TITLE
Let's make a release!

### ISSUE THREAD

**Issue by pjf** [KSP-CKAN/CKAN#302]
**TL;DR at end.**
## The Problem

Last night I lay awake in bed, worrying that we're never going to make a 1.0 release of the CKAN. There was an original and [very brief](https://github.com/KSP-CKAN/CKAN/issues/12) checklist of points to work towards this, and then [a milestone](https://github.com/KSP-CKAN/CKAN/milestones/v1.00%20-%20Usable%20Release) to do the same. However we keep finding new things that need to go into a 1.0 release, and so I fear we're never going to get there.

One of the reasons for this is that our model for software development has but a single branch (`master`), which receives both features, and bug-fixes. The result is that we're almost never in a state of high-stability, because as we know features introduce bugs, and sometimes bug-fixes introduce bugs as well. I've never felt that everything has been stable enough to ship it out the door.

While I appreciate the simplicity that a single-branch development model offers, it almost necessitates calling a feature-freeze in preparation of any release, and results in what I feel is an unnatural level of instability after any given release, as everyone rushes to get their features merged. 
## A Solution

I would like to make not just one, but _two_ releases soon, both a "stable" and a "beta/testing" release. This distinction is an important one; as software that provides strong guarantees with regards to consistency, it's important our stable release feels _really_ stable. The presence of a beta release makes it clear that users _can_ use the latest features, but with the understanding they might not yet be fully tested and stabilised.

To facilitate this, I'm going to soon be creating a v1.0.0 release branch. The _only_ things that can be merged into that branch will be bug-fixes, and features where the absence of which would be considered a bug ( #229 might be the only example of this).

I intend to be working solidly on `v1.0.0` features until release.
### First release vision

First releases are always a big deal, and so I'd like to have two landmark releases made at the same time:
- `v1.0.0` aka _Galaxy Rise_ will consist only of the command-line client, giving us the shortest possible route to an official, stable release.
- `v1.1.0` aka `your suggestion goes here` will include the GUI functionality. By marking it as "unstable" we acknowledge that bugs may still be present, but still maximise the features available to our users.

My hope is that `v1.2.0` will have the GUI formally stabilised, and I imagine the influx of users and developers by having an official release will greatly aid that.
### Versioning

Like many software projects, we'll be using an odd/even notation to denote unstable/stable releases. The current `master` branch will continue to have pre-releases made from it, but they'll be numbered starting with `v1.1.0`. In time, we'll make a new branch off master to form `v1.2.0`, our next stable release incorporating new features, and our development (master) branch will become `v1.3.0`.

While this means that `v1.3.0` will have the odd property of coming out _before_ `v1.2.0`, it keeps consistency that everything in `v1.2.0` is also contained in `v1.3.0`.
### Merge criteria

The criteria for merging to our development (`master`) branch remains as before. The code should be working, tested, and undergone review by another human.

The same applies for release branches, but the change _must_ be a bugfix.

It's encouraged (but not required) you make only a single pull-request (and hence branch from) the most stable branch that can accept it. We'll regularly merge more stable branches into less stable ones; this ensures we don't end up with regressions on new releases.

If unsure, branch from and make pull-requests to `master`; we can always cherry-pick, rebase, or otherwise git-fu them to where they're needed, and this minimises the barrier to new contributors.
### Release cycle

I'm fond of very short release cycles, as no doubt is evidenced by the CKAN pre-releases being made multiple times per day. My sole criteria for making unstable releases is simply "is it better than it was before?". Likewise, I hope to be making new release branches almost as soon as the previous release is stabilised (or in some cases, even sooner).

By minimising the number of changes between each release, we minimise the amount of effort required to ensure those changes are well-tested and stable. It also makes it easier to shift the work of releases between contributors, which is vital for the longevity of the project.
### Further reading

The model I'm proposing is very close to the [gitflow workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) in the Atlassian tutorials.
### Discussion and questions

Highly encouraged. Please comment below. :)

**TL;DR: I'm going to make a release branch shortly, which will accept bugfixes only. We'll be making a simultaneous `v1.0.0` stable (command-line only) and `v1.1.0` unstable (everything!) releases.**
